Removing unnecessary workflow index data from rebuild

Alaschinger opened the pull request:
Instead of reading the whole workflow instance from database when rebuilding the elasticsearch index only four attributes will be used:
- workflow id (long)
- state (int)
- mediapackage id (String)
- organization id (String

They are stored within a new entity object "WorkflowIndexData".

To reduce the database access time the WorkflowIndexData is only sorted by date_created and not filtered within the sql query.
The latest workflows are selected by mapping the date-sorted WorkflowIndexData with the mediapackage id as key value within the repopulate function.

Some attributes and function calls which added additional metadata while rebuilding the workflow index have been removed from the workflowServiceImpl.

Alaschinger committed:
Removing unnecessary workflow index data from rebuild

Instead of reading the whole workflow instance from database when rebuilding the elasticsearch index only four attributes will be used:
- workflow id (long)
- state (int)
- mediapackage id (String)
- organization id (String

They are stored within a new entity object "WorkflowIndexData".

To reduce the database access time the WorkflowIndexData is only sorted by date_created and not filtered within the sql query.
The latest workflows are selected by mapping the date-sorted WorkflowIndexData with the mediapackage id as key value within the repopulate function.

Some attributes and function calls which added additional metadata while rebuilding the workflow index gave been removed from the workflowServiceImpl.

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2296,39 +2271,24 @@ private void removeWorkflowInstanceFromIndex(long workflowInstanceId, Elasticsea
             index.getIndexName());
     Function<Optional<Event>, Optional<Event>> updateFunction = (Optional<Event> eventOpt) -> {
       Event event = eventOpt.orElse(new Event(eventId, orgId));
-      event.setCreator(user.getName());
       event.setWorkflowId(workflowInstanceId);
-      event.setWorkflowDefinitionId(workflowInstance.getTemplate());
```
Is it okay to not set these two?

Alaschinger replied:
Well as far as we know it is. The future will show if anything unexpected goes wrong.

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2156,75 +2153,53 @@ public Map<String, Map<String, String>> getWorkflowStateMappings() {
-            accessControlList = authorizationService.getActiveAcl(instance.getMediaPackage()).getA();
-          }
-
           SecurityUtil.runAs(securityService, organization,
```
If we don't pass modify the creator, I don't think we need to change the security context.

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2156,75 +2153,53 @@ public Map<String, Map<String, String>> getWorkflowStateMappings() {
           Organization organization = null;
           try {
-            organization = organizationDirectoryService.getOrganization(instance.getOrganizationId());
+            organization = organizationDirectoryService.getOrganization(data.getOrganizationId());
           } catch (NotFoundException e) {
-            logger.error("Found workflow with non-existing organization {}", instance.getOrganizationId());
+            logger.error("Found workflow with non-existing organization {}", data.getOrganizationId());
             continue;
           }
```
If we don't need to switch the security context below, we also wouldn't need to retrieve the organization here but could just pass on the organization identifier.

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2156,75 +2153,53 @@ public Map<String, Map<String, String>> getWorkflowStateMappings() {
+        throw new IndexRebuildException(index.getIndexName(), getService(), e);
+      }
+
+      Map<String, WorkflowIndexData> latestWorkflows = new HashMap<>();
+      for (WorkflowIndexData data : workflowIndexData) {
+        latestWorkflows.put(data.getMediaPackageId(), data);
+      }
```
This works assuming the newest workflow for a given media package comes last.
We could be smarter about it if we reverse the order and then only act whenever we get a new media package id without storing this in a map, overwriting the values again and again.

```
mpid  id
a     5
a     1
b     2
c     4
c     3
```

and the something like:

```py
last_mp = ""
for mp, id in data:
  if mp == last_mp:
    continue
  last_mp = mp
  update_index(data)
```

the SQL query for this would ge something like:

```sql
select id, mediapackage_id, … from oc_workflow order by mediapackage_id, id desc;
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowIndexData.java:
```diff
@@ -0,0 +1,99 @@
+@NamedNativeQueries({
+        @NamedNativeQuery(
+                name = "WorkflowIndexData.getAll",
+                query = "SELECT id, state, mediapackage_id, organization_id FROM oc_workflow ORDER BY date_created ASC",
```
Maybe instead something like:
```sql
select id, mediapackage_id, … from oc_workflow order by mediapackage_id, id desc;
```
See comment on code…

Alaschinger committed:
Improved workflow data processing

Workflow data is now processed when doing a rebuild without creating a new map and overwriting the old workflows within it until the latest workflows are fetched. Instead of the data is sorted by mp ID and wf ID and only the first (-> latest) instance of each mp ID will be updated. Also removed the step for changing the security context because we don't need it any longer within the new procedure.

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowIndexData.java:
```diff
@@ -0,0 +1,99 @@
+                        fields = {
+                                  @FieldResult(name = "id",column = "id"),
+                                  @FieldResult(name = "state", column = "state"),
+                                  @FieldResult(name = "mediaPackageId", column = "mediaPackage_id"),
```
Shouldn't matter on most databases, but…
```suggestion
                                  @FieldResult(name = "mediaPackageId", column = "mediapackage_id"),
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabase.java:
```diff
@@ -78,19 +78,19 @@ List<WorkflowInstance> getWorkflowInstancesForCleanup(WorkflowInstance.WorkflowS
-  /**
-   * Gets latest workflow instances for all events.
-   * Selects workflows which haven't been finished (DateCompleted IS NULL) or greatest DateCompleted
+    /**
```
Indentation
```suggestion
  /**
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabase.java:
```diff
@@ -78,19 +78,19 @@ List<WorkflowInstance> getWorkflowInstancesForCleanup(WorkflowInstance.WorkflowS
-   *          only return workflows from this point onwards
-   * @return list of all {@link WorkflowInstance}s
+   *          only return data from this point onwards
+   * @return list of all {@link WorkflowIndexData}s
```
```suggestion
   * @return list of {@link WorkflowIndexData}s
```

lkiesow commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowServiceDatabaseImpl.java:
```diff
@@ -200,22 +200,17 @@ public long countWorkflows(WorkflowInstance.WorkflowState state) throws Workflow
-      query.setMaxResults(limit);
-      logger.debug("Requesting latest workflows using query: {}", query);
-      return query.getResultList();
+      var nativeQuery = em.createNamedQuery("WorkflowIndexData.getAll",WorkflowIndexData.class);
```
```suggestion
      var nativeQuery = em.createNamedQuery("WorkflowIndexData.getAll", WorkflowIndexData.class);
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1260,8 +1250,14 @@ public void update(final WorkflowInstance workflowInstance) throws WorkflowExcep
+          long id = workflowInstance.getId();
+          int state = workflowInstance.getState().ordinal();
+          String mpId = workflowInstance.getMediaPackage().getIdentifier().toString();
+          String orgId = workflowInstance.getOrganizationId().toString();
```
Isn't this a string already?
```suggestion
          String orgId = workflowInstance.getOrganizationId();
```

Alaschinger replied:
Yes, it is. This type conversion was necessary in an earlier version of the code and should be removed now.

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1260,8 +1250,14 @@ public void update(final WorkflowInstance workflowInstance) throws WorkflowExcep
+          String mpId = workflowInstance.getMediaPackage().getIdentifier().toString();
+          String orgId = workflowInstance.getOrganizationId().toString();
+
+          WorkflowIndexData latestWorkflow = new WorkflowIndexData(id, state, mpId, orgId);
+          updateWorkflowInstanceInIndex(latestWorkflow, elasticsearchIndex);
```
Does it make sense to pass the four values to `updateWorkflowInstanceInIndex(…)` directly instead of wrapping this in a new Java object? The first thing which happens in that function is that all values are unpacked again. We wouldn't need to do that if we pass them on as arguments in the first place.

Alaschinger replied:
That's right. I have changed `updateWorkflowInstanceInIndex(…)` now, so that the values don't need to be wrapped and then unwrapped in an object anymore.

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2163,68 +2159,44 @@ public void repopulate(final ElasticsearchIndex index) throws IndexRebuildExcept
+          if (currentMediapackageId.equals(lastMediapackageId)) {
             continue;
           }
+          else {
```
Else is unnecessary since you already terminate the loop with continue. Else invert the confition and do not use continue.
```suggestion
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2163,68 +2159,44 @@ public void repopulate(final ElasticsearchIndex index) throws IndexRebuildExcept
-                  });
-          logIndexRebuildProgress(logger.getSlf4jLogger(), index.getIndexName(), total, current);
         }
       } while (current < total);
```
I just noticed that this is dangerous since new workflows could be added while this is running.
A better termination criteria would be something like:
```suggestion
      } while (workflowIndexData.size() <= 0);
```

lkiesow commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2296,39 +2268,24 @@ private void removeWorkflowInstanceFromIndex(long workflowInstanceId, Elasticsea
+    final long workflowInstanceId = latestWorkflow.getId();
+    final WorkflowState workflowState = WorkflowState.values()[latestWorkflow.getState()];
+    final String eventId = latestWorkflow.getMediaPackageId();
     final String orgId = securityService.getOrganization().getId();
```
We removed `runAs(…)`. Therefore, this is no longer correct. Get the organization from `latestWorkflow`.

lkiesow reviewed (changes requested):
Found mostly minor things…

lkiesow commented:
Given 500 events and 1000 workflows, with this we drop from the 50% rebuild time we got from #3775 to only about 32% of the original time of the 12.x baseline. This speedup now also should be less affected by the ratio of workflows to events, but should be a mostly unconditional speedup:

![rebuild](https://user-images.githubusercontent.com/1008395/172496272-350ea9ae-7d1f-47af-84ee-87af91277695.png)

So, this looks very promising. I still want to do a final code review, though.

It would also be interesting to see how much difference there is if no workflows can be filtered at all. It's likely the 18% difference, but would be interesting to confirm that with a dataset of 500 events and one workflow per event.
